### PR TITLE
[Content Trust] Edge agent handles content trust twin properties

### PR DESCRIPTION
**Pull request by ggjjj**
The design calls out for four fields of content trust information.

    1. path to the root.json file
    2. root ID
    3. path to the root CA file
    4. flag to disableTOFU - Trust On First Use

These fields are added as a part of DockerConfig and class is updated to handle the processing of manifest JSON object to module config. The same changes are made in DockerReportedConfig and the class handles the processing of module config to JSON object which gets reported to IoT hub.

Changes are made in CombinedDockerConfig which packages all info of each module's configuration so that its visible to edgelet.

### CODE REVIEW

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs [hunk 1]
```diff
@@ -20,21 +20,15 @@ public class DockerConfig : IEquatable<DockerConfig>
...
         }
 
-        public DockerConfig(string image, string createOptions)
-            : this(ValidateAndGetImage(image), GetCreateOptions(createOptions))
+        public DockerConfig(string image, string createOptions, ContentTrust contentTrust)
+            : this(ValidateAndGetImage(image), GetCreateOptions(createOptions), Option.Maybe(contentTrust))
         {
         }
 
-        public DockerConfig(string image, CreateContainerParameters createOptions)
```
Comment (CindyXing) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:32]: Would there be any side effect by removing this function?

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs [hunk 2]
```diff
@@ -20,18 +20,18 @@ public class DockerConfigTest
...
+        static readonly DockerConfig Config9 = new DockerConfig("image1:42", @"{""Env"": [""k1=v1"", ""k2=v2"", ""k3=v3""], ""HostConfig"": {""PortBindings"": {""43/udp"": [{""HostPort"": ""43""}], ""42/tcp"": [{""HostPort"": ""42""}]}}}", null);
+
+        static readonly DockerConfig Config10 = new DockerConfig("image1:42", @"{""Env"": [""k11=v11"", ""k22=v22""], ""HostConfig"": {""PortBindings"": {""43/udp"": [{""HostPort"": ""43""}]}}}", null);
+        static readonly DockerConfig Config11 = new DockerConfig("image1:42", @"{""Env"": [""k33=v33"", ""k44=v44""], ""HostConfig"": {""PortBindings"": {""43/udp"": [{""HostPort"": ""43""}]}}}", null);
+        static readonly DockerConfig Config12 = new DockerConfig("image1:42", string.Empty, null);
+        static readonly DockerConfig Config13 = new DockerConfig("image1:42", "{}", null);
+        static readonly DockerConfig Config14 = new DockerConfig("image1:42", "null", null);
+        static readonly DockerConfig Config15 = new DockerConfig("image1:42", "  ", null);
```
Comment (CindyXing) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs:35]: Can we add tests with none or a real DockerContentTrust as input?

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerModuleTest.cs [hunk 3]
```diff
@@ -476,7 +476,7 @@ public void TestSerializeContainerCreateConfig()
         {
             // Arrange
             string createOptions = @"{""Env"": [""k1=v1"",""k2=v2""]}";
-            var config = new DockerConfig("ubuntu:latest", createOptions);
+            var config = new DockerConfig("ubuntu:latest", createOptions, null);
```
Comment (CindyXing) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerModuleTest.cs:479]: Besides null, can we add tests to double check none and a real content trust case?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs (hunk 1 above)
Comment (ggjjj) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:32]: Its a chained constructor. First we get the config as string and then we convert to CreateContainerParameters. So we need both

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs (hunk 2 above)
Comment (ggjjj) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs:35]: I have a test case below TestDeserializationForContentTrustest

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerModuleTest.cs (hunk 3 above)
Comment (ggjjj) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerModuleTest.cs:479]: Good point. i am right now exploring it

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs (hunk 2 above)
Comment (ggjjj) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs:35]: Good point. I will add one for None!

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs [hunk 4]
```diff
@@ -181,7 +186,15 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
                     .Select(token => token?.ToString() ?? string.Empty)
```
Comment (darobs) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:186]: Add to WriteJson:
```
                dockerconfig.ContentTrust.foreach( ct => {
                    writer.WritePropertyName("contentTrust");
                    serializer.Serialize(writer, ct);
                });
```

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerEnvironment.cs [hunk 5]
```diff
@@ -74,7 +74,7 @@ public async Task<ModuleSet> GetModulesAsync(CancellationToken token)
                     : moduleRuntimeInfo.ModuleStatus;
 
                 string image = !string.IsNullOrWhiteSpace(dockerRuntimeInfo.Config.Image) ? dockerRuntimeInfo.Config.Image : dockerModule.Config.Image;
-                var dockerReportedConfig = new DockerReportedConfig(image, dockerModule.Config.CreateOptions, dockerRuntimeInfo.Config.ImageHash);
+                var dockerReportedConfig = new DockerReportedConfig(image, dockerModule.Config.CreateOptions, dockerRuntimeInfo.Config.ImageHash, null);
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerEnvironment.cs:77]: Why isn't dockerRuntimeInfo.NotaryContentTrust not used here? I think the EdgeAgent and EdgeHub specific tests in your PR would catch this if you passed in some value for them.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/RuntimeInfoProvider.cs [hunk 6]
```diff
@@ -116,7 +116,7 @@ string image
             // Figure out module stats and runtime status
             ModuleStatus runtimeStatus = ToRuntimeStatus(inspectResponse.State);
 
-            var reportedConfig = new DockerReportedConfig(image, string.Empty, imageHash);
+            var reportedConfig = new DockerReportedConfig(image, string.Empty, imageHash, null);
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/RuntimeInfoProvider.cs:119]: Shouldn't we report the contentTrust info here?

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.E2E.Test/AgentTests.cs [hunk 7]
```diff
@@ -107,7 +107,7 @@ public async Task AgentStartsUpModules(TestConfig testConfig)
 
                 // Initialize docker configuration for this module.
                 DockerConfig dockerConfig = testConfig.ImageCreateOptions != null
-                    ? new DockerConfig(testConfig.Image, testConfig.ImageCreateOptions)
+                    ? new DockerConfig(testConfig.Image, testConfig.ImageCreateOptions, null)
```
Comment (massand) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.E2E.Test/AgentTests.cs:110]: This maybe addressed in future PR: I am sure you can tell we need to create a Notary server of our own here and add a test case here for agent with notary enabled. It will be good to test with 3 trusted modules E2E  - edgeAgent, edgeHub and simulatedTempSensor.

**Comment by massand** [comment]
Also, one general comment - don't forgot to squash your commits and merge to master. It would be nice to have them squashed before the PR too, so it is easier to merge with PR history.

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs [hunk 8]
```diff
@@ -145,12 +146,100 @@ public void TestSerialization()
...
+                RootID = "45634232423523232423423525253252532523432333",
+                RootCertificatePath = "/home/path/module/rootca",
+                DisableTOFU = true
+            };
+
+            var config = new DockerConfig("image1:42", createOptions, contentTrust);
+            var json = JsonConvert.SerializeObject(config);
+            var expected = "{\"image\":\"image1:42\",\"createOptions\":\"{}\",\"contentTrust\":{\"\":\"\",\"rootID\":\"\",\"rootCertificatePath\":\"/home/path/module/rootca\",\"\":true}}";
+
+            Assert.NotEqual(expected, json);
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs:192]: My idea was to make sure they are not equal to test the error condition. any better way to do it?

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/ContentTrust.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Agent.Docker
+{
+    using System;
+    using Newtonsoft.Json;
+
+    public class ContentTrust : IEquatable<ContentTrust>
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/ContentTrust.cs:7]: Lets go with DockerContentTrust?

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs [hunk 10]
```diff
@@ -5,6 +5,7 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Docker.Test
     using System.Collections.Generic;
     using System.Diagnostics.CodeAnalysis;
     using System.Linq;
+    using Akka.Actor;
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs:8]: Removed it

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs [hunk 11]
```diff
@@ -181,7 +190,15 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
                     .Join();
 
-                return new DockerConfig(jTokenImage?.ToString(), options);
+                var result = obj.TryGetValue("notartyContentTrust", StringComparison.OrdinalIgnoreCase, out JToken jTokenConentTrust);
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:193]: nit: spelling ContentTrust

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerReportedConfig.cs [hunk 12]
```diff
@@ -99,7 +104,15 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
                     .Select(token => token?.ToString() ?? string.Empty)
                     .Join();
 
-                return new DockerReportedConfig(jTokenImage?.ToString(), options, jTokenImageHash?.ToString());
+                var result = obj.TryGetValue("notartyContentTrust", StringComparison.OrdinalIgnoreCase, out JToken jTokenConentTrust);
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerReportedConfig.cs:107]: nit: spelling

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerReportedConfig.cs [hunk 13]
```diff
@@ -99,7 +104,15 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+                if (result)
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerReportedConfig.cs:108]: nit on style: We can inline the TryGetValue here.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet.Docker/CombinedEdgeletConfigProvider.cs [hunk 14]
```diff
@@ -34,7 +34,7 @@ public override CombinedDockerConfig GetCombinedConfig(IModule module, IRuntimeI
             this.MountSockets(module, createOptions);
             this.InjectNetworkAliases(module, createOptions);
 
-            return new CombinedDockerConfig(combinedConfig.Image, createOptions, combinedConfig.AuthConfig);
+            return new CombinedDockerConfig(combinedConfig.Image, createOptions, combinedConfig.NotartyContentTrust, combinedConfig.AuthConfig);
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet.Docker/CombinedEdgeletConfigProvider.cs:37]: Nit: spelling

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs [hunk 15]
```diff
@@ -59,14 +59,14 @@ public async Task<ModuleSet> GetModulesAsync(CancellationToken token)
...
+                    dockerModule = new DockerModule(dockerRuntimeInfo.Name, string.Empty, ModuleStatus.Unknown, Core.RestartPolicy.Unknown, new DockerConfig(Constants.UnknownImage, new CreateContainerParameters(), Option.None<NotaryContentTrust>()), ImagePullPolicy.OnCreate, Core.Constants.HighestPriority, new ConfigurationInfo(), null);
                 }
 
                 Option<ModuleState> moduleStateOption = await this.moduleStateStore.Get(moduleRuntimeInfo.Name);
                 ModuleState moduleState = moduleStateOption.GetOrElse(new ModuleState(0, moduleRuntimeInfo.ExitTime.GetOrElse(DateTime.MinValue)));
 
                 string image = !string.IsNullOrWhiteSpace(dockerRuntimeInfo.Config.Image) ? dockerRuntimeInfo.Config.Image : dockerModule.Config.Image;
-                var dockerReportedConfig = new DockerReportedConfig(image, dockerModule.Config.CreateOptions, dockerRuntimeInfo.Config.ImageHash);
+                var dockerReportedConfig = new DockerReportedConfig(image, dockerModule.Config.CreateOptions, dockerRuntimeInfo.Config.ImageHash, dockerModule.Config.NotartyContentTrust);
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubernetesEnvironment.cs:69]: Nit: spelling

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/CombinedDockerConfig.cs [hunk 16]
```diff
@@ -9,10 +9,11 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Docker
 
     public class CombinedDockerConfig
     {
-        public CombinedDockerConfig(string image, CreateContainerParameters createOptions, Option<AuthConfig> authConfig)
+        public CombinedDockerConfig(string image, CreateContainerParameters createOptions, Option<NotaryContentTrust> notartyContentTrust, Option<AuthConfig> authConfig)
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/CombinedDockerConfig.cs:12]: Nit: spelling

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/CombinedDockerConfig.cs [hunk 17]
```diff
@@ -22,6 +23,9 @@ public CombinedDockerConfig(string image, CreateContainerParameters createOption
         [JsonProperty(Required = Required.AllowNull, PropertyName = "createOptions")]
         public CreateContainerParameters CreateOptions { get; }
 
+        [JsonProperty(Required = Required.Default, PropertyName = "notartyContentTrust")]
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/CombinedDockerConfig.cs:26]: Nit: spelling

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs [hunk 18]
```diff
@@ -20,19 +20,20 @@ public class DockerConfig : IEquatable<DockerConfig>
...
 
-        public DockerConfig(string image, string createOptions)
-            : this(ValidateAndGetImage(image), GetCreateOptions(createOptions))
+        public DockerConfig(string image, string createOptions, Option<NotaryContentTrust> notartyContentTrust)
+            : this(ValidateAndGetImage(image), GetCreateOptions(createOptions), notartyContentTrust)
         {
         }
 
-        public DockerConfig(string image, CreateContainerParameters createOptions)
+        public DockerConfig(string image, CreateContainerParameters createOptions, Option<NotaryContentTrust> notartyContentTrust)
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:32]: Nit: spelling

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs [hunk 19]
```diff
@@ -41,6 +42,8 @@ public DockerConfig(string image, CreateContainerParameters createOptions)
         // https://docs.docker.com/engine/api/v1.25/#operation/ContainerCreate
         public CreateContainerParameters CreateOptions => JsonConvert.DeserializeObject<CreateContainerParameters>(JsonConvert.SerializeObject(this.createOptions));
 
+        public Option<NotaryContentTrust> NotartyContentTrust { get; }
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:45]: Nit: spelling

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs [hunk 20]
```diff
@@ -166,6 +170,11 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
                     writer.WriteValue(chunk);
                 }
 
+                dockerconfig.NotartyContentTrust.ForEach(ct =>
+                    {
+                    writer.WritePropertyName("notartyContentTrust");
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:175]: Nit: Spelling

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerModuleTest.cs [hunk 21]
```diff
@@ -7,6 +7,7 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Docker.Test
     using System.Linq;
     using Microsoft.Azure.Devices.Edge.Agent.Core;
     using Microsoft.Azure.Devices.Edge.Agent.Core.Serde;
+    using Microsoft.Azure.Devices.Edge.Util;
```
[the remark this answers is not in the document]
Comment (ggjjj) [reply edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerModuleTest.cs:10]: It is used to specify Option.None for DockerConfig

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs (hunk 11 above)
Comment (ggjjj) [reply edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:193]: good catch 👍

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs [hunk 22]
```diff
@@ -166,6 +170,11 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
                     writer.WriteValue(chunk);
                 }
 
+                dockerconfig.NotaryContentTrust.ForEach(ct =>
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/DockerConfig.cs:173]: This code does not use the JsonConverter attribution to achieve serialization - which would have been easier to read. It is a good opportunity to make things consistent by using [OptionConverter](https://github.com/Azure/iotedge/blob/75d74603664d1f206585ab3473294236b142a011/edge-util/src/Microsoft.Azure.Devices.Edge.Util/json/OptionConverter.cs). See [SystemModules](https://github.com/Azure/iotedge/blob/027a509549a248647ed41ca7fe1dc508771c8123/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/SystemModules.cs) for example.

If you don't land up converting the serialization to use JsonConverters, consider using option.HasValue and always serialize and write property notaryContentTrust.

File: edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/CombinedDockerConfig.cs [hunk 23]
```diff
@@ -22,6 +23,9 @@ public CombinedDockerConfig(string image, CreateContainerParameters createOption
         [JsonProperty(Required = Required.AllowNull, PropertyName = "createOptions")]
         public CreateContainerParameters CreateOptions { get; }
 
+        [JsonProperty(Required = Required.Default, PropertyName = "notaryContentTrust")]
```
Comment (massand) [inline edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Docker/CombinedDockerConfig.cs:26]: We could align this with authConfig pattern below (i.e. Required.AllowNull)

File: edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs [hunk 24]
```diff
@@ -142,15 +143,87 @@ public void TestDeserializationError()
         [Fact]
         public void TestSerialization()
         {
-            var createOptions = @"{""Env"": [""k1=v1"", ""k2=v2"", ""k3=v3""], ""HostConfig"": {""PortBindings"": {""43/udp"": [" +
+            string createOptions = @"{""Env"": [""k1=v1"", ""k2=v2"", ""k3=v3""], ""HostConfig"": {""PortBindings"": {""43/udp"": [" +
```
Comment (massand) [inline edge-agent/test/Microsoft.Azure.Devices.Edge.Agent.Docker.Test/DockerConfigTest.cs:146]: Leaving var is acceptable here because the type is pretty obvious when plain initialization is happening
